docs/rfcs/0004-agent-flow.md

```diff
@@ -0,0 +1,54 @@
+# Agent Flow - Agent Utilizing Components 
+
+* Date: 2022-03-30
+* Author: Matt Durham (@mattdurham)
+* PR: [grafana/agent#1538](https://github.com/grafana/agent/pull/1538)
+* Status: Draft
+
+## Overarching Problem Statement
+
+The Agents configuration and onboarding is difficult to use. Viewing the effect of configuration changes on telemetry data is difficult. Making the configuration simplier, composable and intuitive allieviates these concerns.
+
+
+## Description
+
+Agent Flow is intended to solve real world needs that the Grafana Agent team have idenfified in conversations with users and developers. 
+
+These broadly include 
+
+- Current configuration is complex and opaque
+- Lack of introspection within the agent
+    - Questions about what telemetry data are being sent
+    - Are rules applying correctly?
+    - How does filtering work?
+- Users have been requesting additional capabilites and adding new features is hard due to coupling between systems, some examples include
+    - Remote Write, Different Input Formats
+    - Different output formats
+    - Filtering ala Relabel Configs is complex and hard to figure out when they occur
+- Lack of understanding how telemetry data moves through agent
+    - Other systems use pipeline/extensions to allow users to understand how data moves through the system
+
+# 1. Introduction and Goals 
+
+This design document describes Agent Flow, which is a system for describing the execution and configuration of telemetry data, along with any secondary configuration required to configure the Grafana Agent. 
+
+Agent Flow refers to both the execution, configuration and visual configurator of data flow.
+
+The primary goals of Agent Flow are
```

Comment:
Is the above a goal or solution?

IE my thought

Goal
`Users need a more intuitive system for configuration that satisfies greater understandability and flexibility.` or `Developers need a system that is less coupled so they can speed up development and allow a feature to focus only on its problem domain`

Solution
`Allowing the users to compose components into a custom pipeline allows them to specifically detail the data flow so that it becomes intuitive how telemetry flows through the system and allows the flexibility of reusing components`

Reply:
That's fair. I'll restate my original suggestion as several goals:

> ```
> * Allow users to more easily understand the impact of their configuration
> * Allow users to collect metrics from all running integrations across a 
>   set of agents without installing extra dependencies
> * Allow users to run integrations based on a dynamic environment
> * Eliminate the need for duplication in the configuration file
> ```

Reply:
Are `Allow users to run integrations based on a dynamic environment` and `Eliminate the need for duplication in the configuration file` similar? Especially if we changed it to `Allow users to run components based on a dynamic environment`

Reply:
> Allow users to run components based on a dynamic environment

This works for me, as long the goal covers re-using a set of discovered targets for metrics, logs, traces, and integrations, I think it's good.